### ISSUE TITLE
insights: [spike] improve scoped insights missing data points

### ISSUE THREAD

**Issue by coury-clark** [sourcegraph/sourcegraph#39883]
In Code Insights missing timeseries values are not recorded. That is to say for a repository set `{A, B, C}`, the matches `{A: 5, B: 2}` will save data for `A` and `B` but not for `C`. In the case of tracking migrations this hurts the experience somewhat, because values can never become zero (missing values are not recorded). To make it slightly more confusing, just-in-time insights (live preview) _will_ show zeroes.

As an example (bottom is JIT and top is after saving):
<img width="512" alt="image" src="https://user-images.githubusercontent.com/5090588/182656548-1ec9629d-3a99-4e01-8a98-43eb07446fa4.png">

We don't want (and can't in many cases) to fill out the missing value set with zeroes. For large insights this would be especially difficult / wasteful. Here are some potential ideas to improve this:

1. For _scoped_ insights populate the missing values and save with zeroes.
2. Track the set of queried times (without repo cardinality) and join this set to the result to coalesce 0 in place of nulls

**Comment by sourcegraph-bot-2** [comment]
Heads up @joelkw @felixfbecker @vovakulikov @unclejustin - the "team/code-insights" label was applied to this issue.

**Comment by coury-clark** [comment]
Related to https://github.com/sourcegraph/sourcegraph/issues/40167

**Comment by Joelkw** [comment]
notes from sync: still need an approach – break up into a spike.

**Comment by leonore** [comment]
I'm merging this with #40167 to be easier to track. two points to make: this is valid for global insights as well (but more visible on scoped insights as these give the live preview which works "as expected"), and option 1) proposed above does not work for capture group cases so would not be an option.

**Comment by leonore** [comment]
after spiking this is what I've found:
1. we calculate in memory what the recording times should be and augment the data on that. however we union, i.e. if we have generated recording times {1, 2, 3} and it turns out actual recording times were {1, 2, 3.5} (simplifying so I don’t have to type out dates) we’ll return {1, 2, 3, 3.5} but {3} will be marked as 0 which could be wrong. however it’s possible we would very rarely fall in this case where the dates wouldn’t match.
2. we calculate recording times on historical backfill for new insights and store them in the database. then we update for new recording times as they come in.
        - if we want this to apply to old insights we can add an oob migration to calculate these

2) is preferred as recording timestamps can vary entirely.

the work to solve this is split up as follows:
- https://github.com/sourcegraph/sourcegraph/issues/42343
- https://github.com/sourcegraph/sourcegraph/issues/42345
- https://github.com/sourcegraph/sourcegraph/issues/42346
